static_analyzer/cluster_relations.py

```
@@ -67,6 +67,180 @@ def _qnames_match(a: str, b: str) -> bool:
+    an edge: an endpoint that names no live symbol (``keep_edge``), and a duplicate the correct
+    pair already carries. Measured on one stored analysis, dropping instead of re-filing orphaned
+    25 of 28 real calls — the mis-attribution is worth repairing, the call is not worth losing.
+
+    An edge whose true pair has no relation to move to stays where it is: inventing a relation
+    here would be authoring architecture in a filter, and silently deleting the call would be
+    worse than leaving it mis-filed where a check can still flag it.
+    """
+    by_pair: dict[tuple[str, str], Relation] = {(relation.src_id, relation.dst_id): relation for relation in relations}
+    additions: dict[tuple[str, str], list[RelationEdge]] = {}
+
+    def find_relation_pair_for_edge(edge: RelationEdge) -> tuple[str, str] | None | bool:
```

Ruff (0.16.0)

[warning] `None` not at the end of the type union.

Move `None` to the end of the type union

(RUF036)